src/Features/Core/Portable/SignatureHelp/SignatureContext.cs

```diff
@@ -0,0 +1,148 @@
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SignatureHelp
+{
+    internal sealed class SignatureContext
+    {
+        private readonly List<SignatureHelpItem> _items;
+        private TextSpan _applicableSpan;
+        private SignatureHelpState _state;
+
+        internal IReadOnlyList<SignatureHelpItem> Items
+        {
+            get { return _items; }
+        }
+
+        /// <summary>
+        /// The document that Signature Help was invoked within.
+        /// </summary>
+        public Document Document { get; }
+
+        /// <summary>
+        /// The caret position when Signature Help was triggered.
+        /// </summary>
+        public int Position { get; }
+
+        /// <summary>
+        /// The triggering action that caused Signature Help to be started.
+        /// </summary>
+        public SignatureHelpTrigger Trigger { get; }
+
+        /// <summary>
+        /// The options that Signature Help was started with.
+        /// </summary>
+        public OptionSet Options { get; }
+
+        /// <summary>
+        /// The cancellation token to use for this operation.
+        /// </summary>
+        public CancellationToken CancellationToken { get; }
+
+        public TextSpan ApplicableSpan => _applicableSpan;
+
+        public SignatureHelpState State => _state;
+
+        /// <summary>
+        /// Creates a <see cref="SignatureContext"/> instance.
+        /// </summary>
+        public SignatureContext(
+            Document document,
+            int position,
+            SignatureHelpTrigger trigger,
+            OptionSet options,
+            CancellationToken cancellationToken)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            this.Document = document;
+            this.Position = position;
+            this.Trigger = trigger;
+            this.Options = options;
+            this.CancellationToken = cancellationToken;
+            _items = new List<SignatureHelpItem>();
+        }
+
+        public void AddItem(SignatureHelpItem item)
+        {
+            if (item == null)
+            {
+                throw new ArgumentNullException(nameof(item));
+            }
+
+            _items.Add(item);
+        }
+
+        public void AddItems(IEnumerable<SignatureHelpItem> items)
+        {
+            if (items == null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            foreach (var item in items)
+            {
+                AddItem(item);
+            }
+        }
+
+        public void SetApplicableSpan(TextSpan span)
+        {
+            _applicableSpan = span;
+        }
+
+        public void SetState(SignatureHelpState state)
```

Comment:
I'm not sure yet. It's already in the signature help system and every provider sets it. I'm going to try and get rid of it completely.